### PR TITLE
better hard and soft SPI sharing

### PR DESCRIPTION
**Pull request by Paciente8159**
- better SPI sharing by separation of hardware and software configurations and settings, and better busy state keeping

### CODE REVIEW

File: uCNC/src/modules/softspi.c [hunk 1]
```diff
@@ -190,6 +186,23 @@ void softspi_bulk_xmit(softspi_port_t *port, const uint8_t *out, uint8_t *in, ui
 	}
 }
 
+bool softspi_isbusy(softspi_port_t *port)
+{
+	if (!port)
+	{
+#ifdef MCU_HAS_SPI
+		port->spiconfig.locked = mcu_spi_port.isbusy;
```
Comment (patryk3211) [inline uCNC/src/modules/softspi.c:194]: Looks like this is assigning to a null pointer

**Comment by patryk3211** [comment]
Other than that, everything looks good.

**Comment by patryk3211** [comment]
Actually, I think I found another issue. `softspi_config` has no guard against concurrent access. On an architecture like STM32 where the SPI hardware gets restarted on config this will be problematic. My suggestion would be to make `softspi_config` store the provided configuration but only after `softspi_start` gets called, apply it to the underlying hardware.

File: uCNC/src/modules/softspi.c (hunk 1 above)
Comment (Paciente8159) [reply uCNC/src/modules/softspi.c:194]: Thanks

**Comment by Paciente8159** [comment]
`softspi_config` should not need guards.
The idea with these changes is that each module declares it's _virtual_ SPI port and keeps it's configuration.

_virtual_ SPI port can be created either using the SOFTSPI macro or the new HARDSPI macro.

For example if SD card module and a TFT module both use the same HW SPI port each creates and HARDSPI port.
That way each as it's own softspi_config but the underlying HW SPI port function calls are the same.

**Comment by patryk3211** [comment]
That's all fine and won't cause any issues, but what if:
1. By calling `softspi_start()` Module A's SPI configuration gets applied.
2. Module A starts a bulk transmission.
3. Module B 'do tasks' handler gets called.
4. Module B decides to change the SPI configuration of its port instance and runs the `softspi_config()` function
5. **`softspi_config()` runs `mcu_spi_config()` and restarts the SPI hardware while Module A is transmitting.**
6. Module B runs `softspi_start()` and waits...

That's the situation that I believe could cause issues.

**Review by patryk3211 — APPROVED** [review]

**Comment by Paciente8159** [comment]
`softspi_config` is to be used at module initialization at most to set an initial configuration.
Otherwise `softspi_start` and `softspi_stop` are the functions to be called.

If you need to change a module configuration that should be done directly in the `softspi_config_t` like it's done in the sd_card module after the SD card initial procedure @100KHz that is then pushed to @10MHz speeds.

The whole way modules/events is implemented has some limitations.
I may add a global locking mechanism at the module level for shared resources but I'm still thing about that.

I'm still running tests to try to figure out the real implications and eventual solutions.

**Pull request merged**